projects/batch-processing/main.go

```diff
@@ -14,45 +24,243 @@ type Converter struct {
 	}
+	log.Println("output CSV file created successfully")
 
-	// Do the conversion!
-	err := c.Grayscale(*inputFilepath, *outputFilepath)
+	// Create a CSV file with the failed records
+	if len(failedRecords) == 1 {
+		log.Println("no failed records found")
+		os.Exit(0)
+	}
+
+	log.Println("Creating failed CSV file ... ")
+	failedPath := fmt.Sprintf("./outputs/%s", *failedFilepath)
+	_, err = CreateCSVFile(failedPath, failedRecords)
 	if err != nil {
-		log.Printf("error: %v\n", err)
+		log.Printf("error creating failed csv file: %v\n", err)
+		os.Exit(1)
 	}
 
-	// Log what we did
 	log.Printf("processed: %q to %q\n", *inputFilepath, *outputFilepath)
 }
+
+func ReadCSV(filepath string) (records [][]string, err error) {
+	// Open the file
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	// Read the CSV file
+	r := csv.NewReader(f)
+	records, err = r.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(records) == 0 {
+		return nil, fmt.Errorf("no records found in the csv file")
+	}
+
+	if len(records[0]) > 1 {
+		return records, fmt.Errorf("more than one column is found in the csv file")
+	}
+
+	return records, nil
+}
+
+func DownloadImage(url string) (string, error) {
+	// Create empty file
+	out, err := os.CreateTemp("", "img-*.jpg")
+	if err != nil {
+		return "", err
+	}
+	defer out.Close()
+	filename := out.Name()
+	// Get the image
+	resp, err := http.Get(url)
+	if err != nil {
+		return filename, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return filename, fmt.Errorf("failed to get the image: %v", resp.Status)
+	}
+
+	// Check if it's an image
+	contentType := resp.Header.Get("Content-Type")
+	if !strings.HasPrefix(contentType, "image/") {
+		return filename, fmt.Errorf("invalid image type: %v", contentType)
+	}
+
+	defer resp.Body.Close()
+
+	// Write the image to the file
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return filename, err
+	}
+	return filename, nil
+}
+
+func (c *Converter) Grayscale(inputFilepath string, outputFilepath string) error {
+	// Convert the image to grayscale using imagemagick
+	// We are directly calling the convert command
+	_, err := c.cmd([]string{
+		"convert", inputFilepath, "-set", "colorspace", "Gray", outputFilepath,
+	})
+	return err
+}
+
+func UploadImage(filename string) (string, error) {
+	// Get the AWS region and role ARN from the environment
+	region := os.Getenv("AWS_REGION")
+	awsRoleArn := os.Getenv("AWS_ROLE_ARN")
+	bucket := os.Getenv("AWS_BUCKET")
+	if region == "" || awsRoleArn == "" {
+		return "", fmt.Errorf("AWS_REGION and AWS_ROLE_ARN environment variables must be set")
+	}
+	// Set up S3 session
+	sess, err := session.NewSession(&aws.Config{
+		Region: aws.String(region),
+	})
+	if err != nil {
+		return "", fmt.Errorf("error creating session: %v\n", err)
+	}
+
+	// Create the credentials from AssumeRoleProvider to assume the role
+	// referenced by the ARN.
+	creds := stscreds.NewCredentials(sess, awsRoleArn)
```

Comment:
You do not need to recreate creds, session, client each time you upload one image. 
You should create these once in a struct constructor and reuse.